src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/entity/Licence.kt

```diff
@@ -146,6 +141,79 @@ abstract class Licence(
     return false
   }
 
+  abstract fun activate(): Licence
+  abstract fun deactivate(): Licence
+  abstract fun submit(submittedBy: CommunityOffenderManager): Licence
+
+  abstract fun updatePrisonInfo(
+    prisonCode: String,
+    prisonDescription: String,
+    prisonTelephone: String?,
+    updatedByUsername: String?,
+  ): Licence
+
+  abstract fun updateAppointmentAddress(appointmentAddress: String?, updatedByUsername: String?): Licence
+  abstract fun updateAppointmentContactNumber(appointmentContact: String?, updatedByUsername: String?): Licence
+  abstract fun updateAppointmentTime(appointmentTime: LocalDateTime, updatedByUsername: String?): Licence
+  abstract fun updateAppointmentPerson(appointmentPerson: String?, updatedByUsername: String?): Licence
+
+  abstract fun updateStatus(
+    statusCode: LicenceStatus,
+    updatedByUsername: String,
+    approvedByUsername: String?,
+    approvedByName: String?,
+    approvedDate: LocalDateTime?,
+    supersededDate: LocalDateTime?,
+    submittedDate: LocalDateTime?,
+    licenceActivatedDate: LocalDateTime?,
+  ): Licence
+
+  abstract fun overrideStatus(
+    statusCode: LicenceStatus,
+    updatedByUsername: String?,
+    licenceActivatedDate: LocalDateTime?,
+  ): Licence
+
+  abstract fun updateConditions(
+    updatedAdditionalConditions: List<AdditionalCondition>? = null,
+    updatedStandardConditions: List<StandardCondition>? = null,
+    updatedBespokeConditions: List<BespokeCondition>? = null,
+    updatedByUsername: String?,
+  ): Licence
+
+  abstract fun updateLicenceDates(
+    status: LicenceStatus? = null,
+    conditionalReleaseDate: LocalDate?,
+    actualReleaseDate: LocalDate?,
+    sentenceStartDate: LocalDate?,
+    sentenceEndDate: LocalDate?,
+    licenceStartDate: LocalDate?,
+    licenceExpiryDate: LocalDate?,
+    topupSupervisionStartDate: LocalDate?,
+    topupSupervisionExpiryDate: LocalDate?,
+    updatedByUsername: String?,
+  ): Licence
+
+  abstract fun updateOffenderDetails(
+    forename: String?,
+    middleNames: String?,
+    surname: String?,
+    dateOfBirth: LocalDate?,
+  ): Licence
+
+  abstract fun updateProbationTeam(
```

Comment:
as there's a lot relying on immutability and copying at the moment (which is a bit of a pain with JPA), all these functions have to be abstract with each subclass implementing the same code but against a subclass specfic subtype copy method. 

i.e: both CRD and variations have similar `updateProbationTeam` methods doing: 
```
) = copy(
    probationAreaCode = probationAreaCode,
    probationAreaDescription = probationAreaDescription,
    probationPduCode = probationPduCode,
    probationPduDescription = probationPduDescription,
    probationLauCode = probationLauCode,
    probationLauDescription = probationLauDescription,
    probationTeamCode = probationTeamCode,
    probationTeamDescription = probationTeamDescription,
  )
```
(`copy` for crd licences returns a `CrdLicence` but returns a `VariationLicence` for variations, we can't call it in a generic way from this subclass).


If we change these fields from vals to vars then we can remove the copy/code duplication which will simplify stuff a lot and remove a lot of code/boiler plate